### ISSUE TITLE
Fails to configure a cluster if database relations are added before the cluster settles

### ISSUE THREAD

**Issue by simskij** [canonical/mysql-k8s-operator#198]
## Steps to reproduce
1. juju deploy mysql-k8s --channel edge -n 3 && juju deploy wordpress-k8s --channel edge && juju relate mysql-k8s:database wordpress-k8s && juju relate mysql-k8s:mysql wordpress-k8s

## Expected behavior

Events to be reconciled or guarded properly to prevent it from blowing up.

## Actual behavior

```
mysql/0       waiting   idle   10.1.14.42         Waiting for instance to join the cluster
mysql/1*      blocked   idle   10.1.14.43         Missing `mysql` relation data
mysql/2       waiting   idle   10.1.14.23         Waiting for instance to join the cluster
```

sometimes resulting in the cluster eventually reconciling after 5+ minutes, sometimes resulting in an `Unable to configure instance` error.

## Versions

Operating system: Ubuntu 22.10

Juju CLI: 3.1.0-genericlinux-amd64

Juju agent: 3.1.0 

Charm revision: latest/edge & latest/edge/cos

microk8s: MicroK8s v1.26.1 revision 4596

## Log output
<details>
<summary> juju debug-log --replay </summary>

```shell
unit-mysql-1: 19:26:22 INFO juju.cmd running containerAgent [3.1.0 924c9e190eb56d9a9324d9cc0dd9dd663c501ac4 gc go1.19.5]
unit-mysql-1: 19:26:22 INFO juju.cmd.containeragent.unit start "unit"
unit-mysql-1: 19:26:22 INFO juju.worker.upgradesteps upgrade steps for 3.1.0 have already been run.
unit-mysql-1: 19:26:22 INFO juju.worker.probehttpserver starting http server on 127.0.0.1:65301
unit-mysql-1: 19:26:22 INFO juju.api connection established to "wss://controller-service.controller-uk8s.svc.cluster.local:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-1: 19:26:22 INFO juju.worker.apicaller [916c47] "unit-mysql-1" successfully connected to "controller-service.controller-uk8s.svc.cluster.local:17070"
unit-mysql-1: 19:26:22 INFO juju.api cannot resolve "controller-service.controller-uk8s.svc.cluster.local": lookup controller-service.controller-uk8s.svc.cluster.local: operation was canceled
unit-mysql-1: 19:26:22 INFO juju.api connection established to "wss://10.152.183.75:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-1: 19:26:22 INFO juju.worker.apicaller [916c47] "unit-mysql-1" successfully connected to "10.152.183.75:17070"
unit-mysql-1: 19:26:22 INFO juju.worker.migrationminion migration phase is now: NONE
unit-mysql-1: 19:26:22 INFO juju.worker.logger logger worker started
unit-mysql-1: 19:26:22 INFO juju.worker.caasupgrader abort check blocked until version event received
unit-mysql-1: 19:26:22 INFO juju.worker.caasupgrader unblocking abort check
unit-mysql-1: 19:26:22 INFO juju.worker.leadership mysql/1 promoted to leadership of mysql
unit-mysql-1: 19:26:22 WARNING juju.worker.proxyupdater unable to set snap core settings [proxy.http= proxy.https= proxy.store=]: exec: "snap": executable file not found in $PATH, output: ""
unit-mysql-1: 19:26:22 INFO juju.agent.tools ensure jujuc symlinks in /var/lib/juju/tools/unit-mysql-1
unit-mysql-1: 19:26:22 INFO juju.worker.uniter unit "mysql/1" started
unit-mysql-1: 19:26:22 INFO juju.worker.uniter resuming charm install
unit-mysql-1: 19:26:22 INFO juju.worker.uniter.charm downloading ch:amd64/jammy/mysql-k8s-57 from API server
unit-mysql-1: 19:26:22 INFO juju.downloader downloading from ch:amd64/jammy/mysql-k8s-57
unit-mysql-1: 19:26:22 INFO juju.downloader download complete ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-1: 19:26:22 INFO juju.downloader download verified ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-2: 19:26:23 INFO juju.cmd running containerAgent [3.1.0 924c9e190eb56d9a9324d9cc0dd9dd663c501ac4 gc go1.19.5]
unit-mysql-2: 19:26:23 INFO juju.cmd.containeragent.unit start "unit"
unit-mysql-2: 19:26:23 INFO juju.worker.upgradesteps upgrade steps for 3.1.0 have already been run.
unit-mysql-2: 19:26:23 INFO juju.worker.probehttpserver starting http server on 127.0.0.1:65301
unit-mysql-0: 19:26:23 INFO juju.cmd running containerAgent [3.1.0 924c9e190eb56d9a9324d9cc0dd9dd663c501ac4 gc go1.19.5]
unit-mysql-0: 19:26:23 INFO juju.cmd.containeragent.unit start "unit"
unit-mysql-0: 19:26:23 INFO juju.worker.upgradesteps upgrade steps for 3.1.0 have already been run.
unit-mysql-0: 19:26:23 INFO juju.worker.probehttpserver starting http server on 127.0.0.1:65301
unit-mysql-0: 19:26:23 INFO juju.api connection established to "wss://controller-service.controller-uk8s.svc.cluster.local:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-0: 19:26:23 INFO juju.worker.apicaller [916c47] "unit-mysql-0" successfully connected to "controller-service.controller-uk8s.svc.cluster.local:17070"
unit-mysql-2: 19:26:23 INFO juju.api connection established to "wss://controller-service.controller-uk8s.svc.cluster.local:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-2: 19:26:23 INFO juju.worker.apicaller [916c47] "unit-mysql-2" successfully connected to "controller-service.controller-uk8s.svc.cluster.local:17070"
unit-mysql-2: 19:26:23 INFO juju.api cannot resolve "controller-service.controller-uk8s.svc.cluster.local": lookup controller-service.controller-uk8s.svc.cluster.local: operation was canceled
unit-mysql-2: 19:26:23 INFO juju.api connection established to "wss://10.152.183.75:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-2: 19:26:23 INFO juju.worker.apicaller [916c47] "unit-mysql-2" successfully connected to "10.152.183.75:17070"
unit-mysql-2: 19:26:23 INFO juju.worker.migrationminion migration phase is now: NONE
unit-mysql-2: 19:26:23 INFO juju.worker.leadership mysql leadership for mysql/2 denied
unit-mysql-2: 19:26:23 INFO juju.worker.logger logger worker started
unit-mysql-2: 19:26:23 WARNING juju.worker.proxyupdater unable to set snap core settings [proxy.http= proxy.https= proxy.store=]: exec: "snap": executable file not found in $PATH, output: ""
unit-mysql-2: 19:26:23 INFO juju.worker.caasupgrader abort check blocked until version event received
unit-mysql-2: 19:26:23 INFO juju.worker.caasupgrader unblocking abort check
unit-mysql-2: 19:26:23 INFO juju.agent.tools ensure jujuc symlinks in /var/lib/juju/tools/unit-mysql-2
unit-mysql-2: 19:26:23 INFO juju.worker.uniter unit "mysql/2" started
unit-mysql-2: 19:26:23 INFO juju.worker.uniter resuming charm install
unit-mysql-2: 19:26:23 INFO juju.worker.uniter.charm downloading ch:amd64/jammy/mysql-k8s-57 from API server
unit-mysql-2: 19:26:23 INFO juju.downloader downloading from ch:amd64/jammy/mysql-k8s-57
unit-mysql-0: 19:26:23 INFO juju.api connection established to "wss://controller-service.controller-uk8s.svc.cluster.local:17070/model/916c4787-18ff-4663-8c90-0dab4550c8bc/api"
unit-mysql-0: 19:26:23 INFO juju.worker.apicaller [916c47] "unit-mysql-0" successfully connected to "controller-service.controller-uk8s.svc.cluster.local:17070"
unit-mysql-0: 19:26:23 INFO juju.worker.migrationminion migration phase is now: NONE
unit-mysql-0: 19:26:23 INFO juju.worker.logger logger worker started
unit-mysql-0: 19:26:23 INFO juju.worker.leadership mysql leadership for mysql/0 denied
unit-mysql-0: 19:26:23 WARNING juju.worker.proxyupdater unable to set snap core settings [proxy.http= proxy.https= proxy.store=]: exec: "snap": executable file not found in $PATH, output: ""
unit-mysql-0: 19:26:23 INFO juju.worker.caasupgrader abort check blocked until version event received
unit-mysql-0: 19:26:23 INFO juju.worker.caasupgrader unblocking abort check
unit-mysql-0: 19:26:23 INFO juju.agent.tools ensure jujuc symlinks in /var/lib/juju/tools/unit-mysql-0
unit-mysql-2: 19:26:23 INFO juju.downloader download complete ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-0: 19:26:23 INFO juju.worker.uniter unit "mysql/0" started
unit-mysql-0: 19:26:23 INFO juju.worker.uniter resuming charm install
unit-mysql-0: 19:26:23 INFO juju.worker.uniter.charm downloading ch:amd64/jammy/mysql-k8s-57 from API server
unit-mysql-0: 19:26:23 INFO juju.downloader downloading from ch:amd64/jammy/mysql-k8s-57
unit-mysql-2: 19:26:23 INFO juju.downloader download verified ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-0: 19:26:24 INFO juju.downloader download complete ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-0: 19:26:24 INFO juju.downloader download verified ("ch:amd64/jammy/mysql-k8s-57")
unit-mysql-1: 19:26:29 INFO juju.worker.uniter hooks are retried true
unit-mysql-1: 19:26:29 INFO juju.worker.uniter found queued "install" hook
unit-mysql-1: 19:26:29 INFO unit.mysql/1.juju-log Running legacy hooks/install.
unit-mysql-1: 19:26:30 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:30 INFO juju.worker.uniter hooks are retried true
unit-mysql-0: 19:26:30 INFO juju.worker.uniter hooks are retried true
unit-mysql-2: 19:26:30 INFO juju.worker.uniter found queued "install" hook
unit-mysql-0: 19:26:30 INFO juju.worker.uniter found queued "install" hook
unit-mysql-1: 19:26:31 INFO juju.worker.uniter.operation ran "database-peers-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:31 INFO unit.mysql/1.juju-log mysql:22: Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-2: 19:26:31 INFO unit.mysql/2.juju-log Running legacy hooks/install.
unit-mysql-0: 19:26:31 INFO unit.mysql/0.juju-log Running legacy hooks/install.
unit-mysql-1: 19:26:31 INFO juju.worker.uniter.operation ran "mysql-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:32 INFO unit.mysql/1.juju-log database:23: Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-2: 19:26:32 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:32 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:32 INFO juju.worker.uniter.operation ran "database-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:32 INFO juju.worker.uniter found queued "leader-elected" hook
unit-mysql-1: 19:26:32 INFO unit.mysql/1.juju-log Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-2: 19:26:32 INFO juju.worker.uniter.operation ran "database-peers-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:32 INFO juju.worker.uniter.operation ran "database-peers-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:33 INFO juju.worker.uniter.operation ran "leader-elected" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:33 INFO unit.mysql/2.juju-log mysql:22: Unit is not leader, nooping `mysql` relation created
unit-mysql-0: 19:26:33 INFO unit.mysql/0.juju-log mysql:22: Unit is not leader, nooping `mysql` relation created
unit-mysql-1: 19:26:33 INFO unit.mysql/1.juju-log Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-2: 19:26:33 INFO juju.worker.uniter.operation ran "mysql-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:33 INFO juju.worker.uniter.operation ran "mysql-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:33 INFO juju.worker.uniter.operation ran "mysql-pebble-ready" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:33 INFO unit.mysql/1.juju-log Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-2: 19:26:33 INFO juju.worker.uniter.operation ran "database-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:34 INFO juju.worker.uniter.operation ran "database-relation-created" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:34 INFO juju.worker.uniter.operation ran "database-storage-attached" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:34 INFO juju.worker.uniter.operation ran "leader-settings-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:34 INFO unit.mysql/1.juju-log Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-0: 19:26:34 INFO juju.worker.uniter.operation ran "leader-settings-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:34 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:34 INFO juju.worker.uniter found queued "start" hook
unit-mysql-0: 19:26:34 INFO unit.mysql/0.juju-log Retrieving the total memory of the server
unit-mysql-2: 19:26:35 INFO juju.worker.uniter.operation ran "mysql-pebble-ready" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:35 INFO unit.mysql/1.juju-log Running legacy hooks/start.
unit-mysql-2: 19:26:35 INFO unit.mysql/2.juju-log Retrieving the total memory of the server
unit-mysql-1: 19:26:35 INFO unit.mysql/1.juju-log Unit not ready to execute `mysql` relation created. Deferring
unit-mysql-1: 19:26:35 INFO unit.mysql/1.juju-log Retrieving the total memory of the server
unit-mysql-0: 19:26:46 INFO unit.mysql/0.juju-log Configuring instance
unit-mysql-2: 19:26:48 INFO unit.mysql/2.juju-log Configuring instance
unit-mysql-1: 19:26:49 INFO unit.mysql/1.juju-log Configuring instance
unit-mysql-0: 19:26:53 INFO juju.worker.uniter.operation ran "mysql-pebble-ready" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:54 INFO juju.worker.uniter.operation ran "database-storage-attached" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:54 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:54 INFO juju.worker.uniter found queued "start" hook
unit-mysql-0: 19:26:55 INFO unit.mysql/0.juju-log Running legacy hooks/start.
unit-mysql-2: 19:26:55 INFO juju.worker.uniter.operation ran "database-storage-attached" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:55 INFO unit.mysql/1.juju-log Creating cluster on the leader unit
unit-mysql-0: 19:26:55 INFO juju.worker.uniter.operation ran "start" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:56 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:56 INFO juju.worker.uniter found queued "start" hook
unit-mysql-0: 19:26:56 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:56 INFO unit.mysql/2.juju-log Running legacy hooks/start.
unit-mysql-0: 19:26:57 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:57 INFO juju.worker.uniter.operation ran "start" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:57 INFO juju.worker.uniter.operation ran "start" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:57 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:57 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:58 WARNING unit.mysql/1.juju-log database-peers:20: DEPRECATION WARNING - `mysql` is a legacy interface
unit-mysql-1: 19:26:58 WARNING unit.mysql/1.juju-log database-peers:20: mysql-interface-user or mysql-interface-database not set
unit-mysql-1: 19:26:58 INFO unit.mysql/1.juju-log database-peers:20: Kubernetes pod label primary created
unit-mysql-1: 19:26:58 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:58 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:58 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:59 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:26:59 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:26:59 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:26:59 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:27:00 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:27:00 INFO unit.mysql/1.juju-log database-peers:20: Kubernetes pod label primary created
unit-mysql-1: 19:27:00 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:27:00 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:27:00 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:27:01 INFO juju.worker.uniter.operation ran "database-peers-relation-joined" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:27:01 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:27:02 INFO unit.mysql/1.juju-log database-peers:20: Kubernetes pod label primary created
unit-mysql-1: 19:27:02 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-mysql-2: 19:31:17 INFO unit.mysql/2.juju-log Unit state is waiting
unit-mysql-2: 19:31:17 INFO unit.mysql/2.juju-log Unit state is waiting
unit-mysql-2: 19:31:17 WARNING unit.mysql/2.juju-log No relation: certificates
unit-mysql-0: 19:31:17 INFO unit.mysql/0.juju-log Unit state is waiting
unit-mysql-0: 19:31:17 INFO unit.mysql/0.juju-log Unit state is waiting
unit-mysql-0: 19:31:17 WARNING unit.mysql/0.juju-log No relation: certificates
unit-mysql-2: 19:31:17 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
unit-mysql-0: 19:31:17 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
unit-mysql-1: 19:32:24 INFO unit.mysql/1.juju-log Unit workload member-state is online with member-role primary
unit-mysql-1: 19:32:24 INFO unit.mysql/1.juju-log Kubernetes pod label primary created
unit-mysql-1: 19:32:24 WARNING unit.mysql/1.juju-log No relation: certificates
unit-mysql-1: 19:32:24 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
```
</details>

## Additional context

The deployment is made using terraform, hence the odd looking reproduction steps - which are an attempt to recreate that.

**Comment by taurus-forever** [comment]
Hi @simskij, Thank you for the report! I think we have several issues here.

1) can you please try the following from your side:
```
juju deploy mysql-k8s --channel 8.0/edge -n 3 \
  --config mysql-interface-user=test-user \
  --config mysql-interface-database=test-database && \
juju deploy wordpress-k8s --channel edge && \
juju relate mysql-k8s:mysql wordpress-k8s
```

* We have recently migrated to track `8.0`, consider to use it instead of `latest` (the `latest` track will be closed).
* The charm `wordpress-k8s` is currently support legacy interface `mysql` only, no need to relate both legacy `mysql` and modern `mysql_client` interface simultaneously (we will add protection here to avoid damage if both relations are requested).
* The legacy interface requires manual set for `user` and `database` (at the moment). Auto-generation [planned](https://warthogs.atlassian.net/browse/DPE-1512).

2) We will describe all interfaces in mysql-k8s readme like we have in [mysql vm charm](https://github.com/canonical/mysql-operator#relations).

3) Unfortunately the blocking message `Missing 'mysql' relation data` is useless as the main information in debug log only:
```
unit-mysql-1: 19:26:58 WARNING unit.mysql/1.juju-log database-peers:20: DEPRECATION WARNING - `mysql` is a legacy interface
unit-mysql-1: 19:26:58 WARNING unit.mysql/1.juju-log database-peers:20: mysql-interface-user or mysql-interface-database not set
```
Will be no longer be topical after [options auto-generation](https://warthogs.atlassian.net/browse/DPE-1512).

Tnx for reporting this issue!

**Comment by paulomach** [comment]
@simskij , trying with the commands from Alex there's it works after some tries, but wp charm errors with a MySQL 2003 (connecton refused) on [this line](https://github.com/canonical/wordpress-k8s-operator/blob/75764761f42fe4b9ff9b76d2f22c22688b8f5292/src/charm.py#L1497) 

While not ideal, I suggest that the wp charm try a db connection and defer on error, what do you think?

The same will happen when using the new interfaces, e.g.:
```
juju deploy mysql-k8s --channel 8.0/edge -n 3 --trust && \
juju deploy wordpress-k8s --channel edge && \
juju relate mysql-k8s:database wordpress-k8s
```

Meanwhile, I'll try to investigate it further - for now I'm suspecting k8s svc endpoint lagging

**Comment by simskij** [comment]
> @simskij , trying with the commands from Alex there's it works after some tries, but wp charm errors with a MySQL 2003 (connecton refused) on [this line](https://github.com/canonical/wordpress-k8s-operator/blob/75764761f42fe4b9ff9b76d2f22c22688b8f5292/src/charm.py#L1497)
> 
> While not ideal, I suggest that the wp charm try a db connection and defer on error, what do you think?
> 
> The same will happen when using the new interfaces, e.g.:
> 
> ```
> juju deploy mysql-k8s --channel 8.0/edge -n 3 --trust && \
> juju deploy wordpress-k8s --channel edge && \
> juju relate mysql-k8s:database wordpress-k8s
> ```
> 
> Meanwhile, I'll try to investigate it further - for now I'm suspecting k8s svc endpoint lagging

I agree. That sounds like a reasonable change to the WP charm

**Comment by weiiwang01** [comment]
I do not prefer deferring events because re-emitting the deferred event depends on the next event, which is not controlled by the charm. Sometimes, it might result in a significant delay. 

Currently, after the database relation is established and database connection information is provided through the relation, the WordPress charm will retry for up to one minute before declaring a database error. What is the typical delay between the database connection information provided through relation and the database being ready? I can adjust the WordPress database probe timeout accordingly. The current timeout is 60 seconds

**Comment by weiiwang01** [comment]
> > @simskij , trying with the commands from Alex there's it works after some tries, but wp charm errors with a MySQL 2003 (connecton refused) on [this line](https://github.com/canonical/wordpress-k8s-operator/blob/75764761f42fe4b9ff9b76d2f22c22688b8f5292/src/charm.py#L1497)
> > While not ideal, I suggest that the wp charm try a db connection and defer on error, what do you think?
> > The same will happen when using the new interfaces, e.g.:
> > ```
> > juju deploy mysql-k8s --channel 8.0/edge -n 3 --trust && \
> > juju deploy wordpress-k8s --channel edge && \
> > juju relate mysql-k8s:database wordpress-k8s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Meanwhile, I'll try to investigate it further - for now I'm suspecting k8s svc endpoint lagging
> 
> I agree. That sounds like a reasonable change to the WP charm. What do you think @weiiwang01 ?

I do not prefer deferring events because re-emitting the deferred event depends on the next event, which is not controlled by the charm. Sometimes, it might result in a significant delay.

Currently, after the database relation is established and database connection information is provided through the relation, the WordPress charm will retry for up to one minute before declaring a database error. What is the typical delay between the database connection information provided through relation and the database being ready? I can adjust the WordPress database probe timeout accordingly. The current timeout is 60 seconds.

**Comment by simskij** [comment]
@weiiwang01 In that case I would probably make it a good five minutes or so before timing out, given how long it can take for the cluster to reconcile (only happens through `update-status` events which means it can potentially take a **while**)

**Comment by weiiwang01** [comment]
@simskij I updated the database probe timeout in the WordPress charm from 1 min to 5 mins.

https://github.com/canonical/wordpress-k8s-operator/pull/70

**Comment by simskij** [comment]
Been discussing this issue back and forth outside of this ticket and since we're relying on update-status to trigger the next stage of the MySQL cluster forming, there really isn't a way to make this faster atm other than making the update-status interval shorter. With the change in the Wordpress charm I think this will be good enough for now, so I'll close this issue.
